wallet, gui: bugfix, getAvailableBalance skips selected coins

furszy opened the pull request:
Fixes https://github.com/bitcoin-core/gui/issues/688 and https://github.com/bitcoin/bitcoin/issues/26687.

First Issue Description (https://github.com/bitcoin-core/gui/issues/688):

The previous behavior for `getAvailableBalance`, when the coin control had selected coins, was to return the sum of them. Instead, we are currently returning the wallet's available total balance minus the selected coins total amount.

Reason:
Missed to update the `GetAvailableBalance` function to include the coin control selected coins on #25685.

Context:
Since #25685 we skip the selected coins inside `AvailableCoins`, the reason is that there is no need to waste resources walking through the entire wallet's txes map just to get coins that could have gotten by just doing a simple `mapWallet.find`).

Places Where This Generates Issues (only when the user manually select coins via coin control):
1) The GUI balance check prior the transaction creation process.
2) The GUI "useAvailableBalance" functionality.

Note 1:
As the GUI uses a balance cache since https://github.com/bitcoin-core/gui/pull/598, this issue does not affect the regular spending process. Only arises when the user manually select coins.

Note 2:
Added test coverage for the `useAvailableBalance` functionality.

----------------------------------

Second Issue Description (https://github.com/bitcoin/bitcoin/issues/26687):

As we are using a cached balance on `WalletModel::getAvailableBalance`,
the function needs to include the watch-only available balance for wallets
with private keys disabled.

furszy committed:
gui: bugfix, getAvailableBalance skips selected coins

The previous behavior for getAvailableBalance when coin control
has selected coins was to return the sum of them. Instead, we
are currently returning the wallet's available total balance minus
the selected coins total amount.

This turns into a GUI-only issue for the "use available balance"
button when the user manually select coins in the send screen.

Reason:
We missed to update the GetAvailableBalance function to include
the coin control selected coins on #25685.

Context:
Since #25685 we skip the selected coins inside `AvailableCoins`,
the reason is that there is no need to traverse the wallet's
txes map just to get coins that can directly be fetched by
their id.

furszy committed:
test: add coverage for 'useAvailableBalance' functionality

The following cases were covered:

Case 1: No coin control selected coins.
  - 'useAvailableBalance' should fill the amount edit box with the total available balance.

Case 2: With coin control selected coins.
  - 'useAvailableBalance' should fill the amount edit box with the sum of the selected coins values.

achow101 commented:
While you're in this function, can you also fix #26687. It's a similar issue - we're now filtering for only spendable coins which excludes watch-only coins.

furszy commented:
> While you're in this function, can you also fix #26687. It's a similar issue - we're now filtering for only spendable coins which excludes watch-only coins.

sure @achow101, will check it. Feel free to cc me for other similar stuff next time.

furszy committed:
gui: 'getAvailableBalance', include watch only balance

Only for wallets with private keys disabled.

The returned amount need to include the watch-only
available balance too.

Solves #26687.

furszy commented:
Added fixes for https://github.com/bitcoin/bitcoin/issues/26687.

1) As we are using a cached balance on `WalletModel::getAvailableBalance`,
the function needs to include into the response the watch-only available
balance on wallets with private keys disabled.

2) Inside the transaction creation process, the wallet needs to include the
non-spendable coins if private keys are disabled.

Note:
Let me know if want me to decouple this two fixes into a new PR.

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -897,7 +897,9 @@ static util::Result<CreatedTransactionResult> CreateTransactionInternal(
     if (coin_control.m_allow_other_inputs) {
-        available_coins = AvailableCoins(wallet, &coin_control, coin_selection_params.m_effective_feerate);
+        CoinFilterParams params; // include non-spendable coins if this is a watch-only wallet
+        params.only_spendable = !wallet.IsWalletFlagSet(WALLET_FLAG_DISABLE_PRIVATE_KEYS);
```
Can we test for this?

furszy replied:
Actually, this change is not needed here. Will just drop it.
I was using it to add support for creating PSBTs on legacy wallets that only know the pubkey hash and not the pubkey (in other terms, use `importaddress` instead of `importpubkey`). Which has never being possible.

furszy committed:
test,gui: decouple chain and wallet initialization from test case

Prepare ground for legacy watch-only test.

furszy committed:
test,gui: add coverage for PSBT creation on legacy watch-only wallets

furszy committed:
test,gui: decouple widgets and model into a MiniGui struct

So it can be reused across tests.

theStack reviewed (commented):
Concept ACK

Saw this problem yesterday while I was experimenting with the GUI and noticed that trying to spend a single available UTXO via coin control failed with a "The amount exceeds your balance." error -- IIUC, this issue appears whenever the balance of the not-selected coins is smaller than the balance of the selected coins.

MarcoFalke commented:
Needs rebase

furszy commented:
rebased, conflicts solved.

Sjors commented:
Tested that this fixes the coin selection + use available balance GUI issue for me. Haven't reviewed the code yet.

furszy commented:
rebased, conflict solved.

Sjors reviewed (commented):
Tested and reviewed 86b43c78ecd5a68b5569b1c2a7b2696b1f70ef21.

Especially happy with all the new tests. Some questions / issues:

furszy commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -791,8 +791,13 @@ void SendCoinsDialog::useAvailableBalance(SendCoinsEntry* entry)
     m_coin_control->fAllowWatchOnly = model->wallet().privateKeysDisabled() && !model->wallet().hasExternalSigner();
 
+    // Same behavior as send: if we have selected coins, only obtain their available balance.
+    // future, introduce a checkbox to customize this value.
```
Basically to follow the same procedure as the send process (`SendCoinsDialog::PrepareSendText`), it's is at line 288-290.

Currently, the GUI disables the automatic selection of "other inputs" when the user manually selected coins (the process will only use those). But, in the future, that could also be customizable. The user might want to let the wallet add new inputs, even when the user had selected some, if there aren't enough funds to cover for the tx target amount.

Sjors replied:
I see. I think it's better to drop that comment and instead open a GUI issue. The checkbox could say "Add more coins as needed" to match the send RPC behaviour. And it can point to the line `!coin_control.HasSelected();` that needs to be changed to implement this.

furszy replied:
np about creating a gui issue for this.

still, I'm not fan of the link to this particular line because this isn't the only place that has to be modified to introduce the feature. Plus, code changes over time, so the issue could easily get behind, which would mean more maintenance burden for us.

furszy replied:
pushed

furszy commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -791,8 +791,13 @@ void SendCoinsDialog::useAvailableBalance(SendCoinsEntry* entry)
+    // Same behavior as send: if we have selected coins, only obtain their available balance.
+    // future, introduce a checkbox to customize this value.
+    CCoinControl coin_control = *m_coin_control;
+    coin_control.m_allow_other_inputs = !coin_control.HasSelected();
```
The `m_allow_other_inputs` flag is different to the `fAllowWatchOnly` flag.

`fAllowWatchOnly` set is actually redundant here as `privateKeysDisabled()` and `hasExternalSigner()` are wallet constants. It can be set only once, when the wallet is set in the GUI and don't be re-set again.

And, at the contraire, the `m_allow_other_inputs` flag can change as it depends on whether the user manually selected coins or not, which could be modified at any time. Thus why I would try to not set the coin control member here.

The best path would be to set `m_allow_other_inputs` at the same moment when the user manually selects/deselect coins in the coin control dialog. But.. I think that it's better to do that in a follow-up to not expand this PR much.

Sjors replied:
Ah, maybe add a comment:

```cpp
// Copy to avoid modifying m_coin_control 
```

furszy replied:
Sure. Doesn't hurt to add a comment.

furszy replied:
pushed

furszy commented on src/wallet/interfaces.cpp:
```diff
@@ -403,7 +404,24 @@ class WalletImpl : public Wallet
+        if (coin_control.HasSelected()) {
+            FastRandomContext rng{};
+            CoinSelectionParams params(rng);
+            // Note: for now, swallow any error.
```
Was remembering why I left it unhandled:
And it was because this cannot happen. `FetchSelectedInputs` fails only if any of the selected coins are not solvable or invalid. And we set the selected coins via the coin control dialog, which is loaded from the wallet's available UTXOs, which are always valid.

Sjors replied:
I see. In that case, just add what you just wrote as a comment.

furszy replied:
Ok, sounds good. Will do it in the next push.

furszy commented:
Feedback tackled. Thanks Sjors.

Only a comment change, [diff](https://github.com/bitcoin/bitcoin/compare/86b43c78ecd5a68b5569b1c2a7b2696b1f70ef21..68eed5df8656bed1be6526b014e58d3123102b03).

Sjors commented:
ACK 68eed5df8656bed1be6526b014e58d3123102b03

Remaining feedback is only about source code comments.

achow101 commented:
ACK 68eed5df8656bed1be6526b014e58d3123102b03

ryanofsky commented on src/wallet/interfaces.cpp:
```diff
@@ -403,7 +404,24 @@ class WalletImpl : public Wallet
     CAmount getBalance() override { return GetBalance(*m_wallet).m_mine_trusted; }
     CAmount getAvailableBalance(const CCoinControl& coin_control) override
     {
-        return GetAvailableBalance(*m_wallet, &coin_control);
```
> [8cd7730](https://github.com/bitcoin/bitcoin/commit/8cd77302305a2387368aabe8744b1859caf9956a): paging @ryanofsky: do you think we should move all this complexity here to the interface?

Right, probably it would be better to move the new code from the `wallet:WalletImpl::getAvailableBalance()` method in `src/wallet/interfaces.cpp` into the `wallet::GetAvailableBalance()` function in `src/wallet/spend.cpp` or into another function there like `wallet::GetSelectedCoinsBalance`. Reasoning is

- `src/wallet/interfaces.cpp` is supposed to define a public interface to wallet code that GUI code can call without accessing wallet internals. It is just supposed to forward calls from the public interface to private functions, and not really designed to contain coin selection code.
- It is easier to write unit tests for standalone `spend.cpp` functions than interface methods that require constructing a larger interface.
- It is easier to share code between RPC and GUI interfaces right now if code is standalone

furszy replied:
Sure. Having a proxy class is nice for any future IPC mechanism. All the protobuf/protocan/etc messages are simpler to craft in this way.

Based on how close we are to v25 branch-off, I'm thinking on whether should do it now or later :/. Otherwise I would push it right away.

Probably, could tackle it on a quick follow-up with https://github.com/bitcoin/bitcoin/pull/26699#discussion_r1156018577 and few other cleanups as well.

theStack commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -403,7 +403,7 @@ void SendCoinsDialog::presentPSBT(PartiallySignedTransaction& psbtx)
     ssTx << psbtx;
     GUIUtil::setClipboard(EncodeBase64(ssTx.str()).c_str());
-    QMessageBox msgBox;
+    QMessageBox msgBox(this);
```
nit: Is there any specific reason for this change? Without it, everything seems to work still fine on my side (but obviously, it shouldn't hurt to construct the message box with a parent...)

furszy replied:
It was because I was accessing the parent to click on the message box from the unit test (the dialog `exec()` method blocks the thread until the user presses any of the buttons). And, later on, changed my mind and accessed it via the `QApplication::topLevelWidgets` call.

It doesn't hurt to have it. This would had been a leak if the message box would had been constructed via a `new` call.

theStack reviewed (approved):
ACK 68eed5df8656bed1be6526b014e58d3123102b03

Reviewed the changeset (with only light review for the qt unit tests, as I'm not too familiar in that area), verified manually that the [GUI coin selection issue](https://github.com/bitcoin/bitcoin/pull/26699#pullrequestreview-1253771673) is fixed, tested the tests by reverting the fix commits dc1cc1c35995dc09085b3d9270c445b7923fdb51 and cd98b717398f7b13ace91ea9efac9ce1e60b4d62 each and checking that the corresponding unit tests fail. LGTM.

hebasto commented on src/qt/test/wallettests.cpp:
```diff
@@ -313,6 +391,76 @@ void TestGUI(interfaces::Node& node)
+    sendCoinsDialog.getCoinControl()->destChange = GetDestinationForKey(test.coinbaseKey.GetPubKey(), OutputType::LEGACY);
+
+    // Time to reject "save" PSBT dialog ('SendCoins' locks the main thread until the dialog receives the event).
+    QTimer timer;
+    timer.setInterval(500);
+    QObject::connect(&timer, &QTimer::timeout, [&](){
+        for (QWidget* widget : QApplication::topLevelWidgets()) {
+            if (widget->inherits("QMessageBox")) {
+                QMessageBox* dialog = qobject_cast<QMessageBox*>(widget);
+                QAbstractButton* button = dialog->button(QMessageBox::Discard);
+                button->setEnabled(true);
+                button->click();
+                timer.stop();
+                break;
+            }
+        }
+    });
+    timer.start(500);
```
It seems all these lines of code might be replaced with `ConfirmMessage(nullptr, 500ms);`.

furszy replied:
> It seems all these lines of code might be replaced with `ConfirmMessage(nullptr, 500ms);`.

That should trigger the filesystem window to dump the psbt info to a file, which would incur in another step to cancel the dialog. The idea here is to obtain the data through the clipboard only (see few lines below), not through a file.

That being said, agree that we could move this into the utility file in the future and cleanup some code.

hebasto replied:
> That will trigger the filesystem window to dump the psbt info to a file...

Why? The `ConfirmMessage` triggers the default button, which is `QMessageBox::Discard`, no?

furszy replied:
> > That will trigger the filesystem window to dump the psbt info to a file...
> 
> Why? The `ConfirmMessage` triggers the default button, which is `QMessageBox::Discard`, no?

Yeah ok, I did not see the manual `setDefaultButton` call. That will work fine while the dialog takes less than 500ms in popping up. If it takes longer, the test will stall because `ConfirmMessage` executes runs the time only once (`QTimer::singleShot`).
